# HG changeset patch
# User Rik <rik@octave.org>
# Date 1545419948 28800
#      Fri Dec 21 11:19:08 2018 -0800
# Branch stable
# Node ID da54f14258e1e80cecde22096c0b6829583a407e
# Parent  01fdb337fa20a4a6de12ff445f50e7434dd625f7
fminsearch.m: Fix sign of results for intermediate reporting (bug #55261).

* fminsearch.m (nmsmax): Multiply all occurences where function value (f) is displayed to the user by the direction of the algorithm "dirn" to present the true value.  For minimization,
dirn is always -1 which reverses the sign of the result.

diff --git a/scripts/optimization/fminsearch.m b/scripts/optimization/fminsearch.m
--- a/scripts/optimization/fminsearch.m
+++ b/scripts/optimization/fminsearch.m
@@ -277,17 +277,17 @@ function [x, exitflag, output] = nmsmax 
   V = [zeros(n,1) eye(n)];
   f = zeros (n+1,1);
   V(:,1) = x0;
   f(1) = dirn * fun (x, varargin{:});
   fmax_old = f(1);
   nf = 1;
 
   if (trace == 1)
-    printf ("f(x0) = %9.4e\n", f(1));
+    printf ("f(x0) = %9.4e\n", dirn * f(1));
   endif
 
   k = 0; m = 0;
 
   ## Set up initial simplex.
   scale = max (norm (x0, Inf), 1);
   if (stopit(4) == 0)
     ## Regular simplex - all edges have same length.
@@ -316,17 +316,17 @@ function [x, exitflag, output] = nmsmax 
   f = f(j);
   V = V(:,j);
 
   exitflag = 0;
 
   if (! isempty (outfcn))
     optimvalues.iteration = 0;
     optimvalues.funccount = nf;
-    optimvalues.fval = f(1);
+    optimvalues.fval = dirn * f(1);
     optimvalues.procedure = how;
     state = "init";
     stop = outfcn (x, optimvalues, state);
     if (stop)
       msg = "Stopped by OutputFcn\n";
       exitflag = -1;
     endif
   endif
@@ -346,17 +346,17 @@ function [x, exitflag, output] = nmsmax 
       if (! isempty (savit))
         x(:) = V(:,1);
         eval (["save " savit " x fmax nf"]);
       endif
     endif
     if (trace == 1)
       printf ("Iter. %2.0f,", k);
       printf ("  how = %-11s", [how ","]);
-      printf ("nf = %3.0f,  f = %9.4e  (%2.1f%%)\n", nf, fmax, ...
+      printf ("nf = %3.0f,  f = %9.4e  (%2.1f%%)\n", nf, dirn * fmax, ...
               100*(fmax-fmax_old)/(abs(fmax_old)+eps));
     endif
     fmax_old = fmax;
 
     ## Three stopping tests from MDSMAX.M
 
     ## Stopping Test 1 - f reached target value?
     if (fmax >= stopit(3))
@@ -383,17 +383,17 @@ function [x, exitflag, output] = nmsmax 
                       size_simplex, tol, step_f, tol_f);
       exitflag = 1;
       break;
     endif
 
     ## Call OutputFcn
     if (! isempty (outfcn))
       optimvalues.funccount = nf;
-      optimvalues.fval = f(1);
+      optimvalues.fval = dirn * f(1);
       optimvalues.iteration = k;
       optimvalues.procedure = how;
       state = "iter";
       stop = outfcn (x, optimvalues, state);
       if (stop)
         msg = "Stopped by OutputFcn\n";
         exitflag = -1;
         break;
@@ -465,17 +465,17 @@ function [x, exitflag, output] = nmsmax 
     printf (msg);
   endif
   x(:) = V(:,1);
 
   ## FIXME: Should outputfcn be called only if exitflag != 0,
   ##        i.e., only when we have successfully converged?
   if (! isempty (outfcn))
     optimvalues.funccount = nf;
-    optimvalues.fval = f(1);
+    optimvalues.fval = dirn * f(1);
     optimvalues.iteration = k;
     optimvalues.procedure = how;
     state = "done";
     outfcn (x, optimvalues, state);
   endif
 
   ## output
   output.iterations = k;
